### PR TITLE
Features upgrade of `Embedding model`

### PR DESCRIPTION
**Pull request by quic-amitraj**
1. Added `Pooling` support to Embedding Model
2. Added multiple seq_len support for the embedding model using `QEffAutoModel`
4. Added test for pooling and multiple seq_len

### CODE REVIEW

File: QEfficient/utils/_utils.py [hunk 1]
```diff
@@ -662,3 +662,9 @@ def filter_kwargs(func, kwargs):
     """
     valid_args = inspect.signature(func).parameters
     return {key: value for key, value in kwargs.items() if key in valid_args}
+
+
+def custom_format_warning(msg, category, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (quic-amitraj) [reply QEfficient/utils/_utils.py:667]: image

I was thinking its a good idea to notify user that pooling is appended to the model. We will add this to the logger as well but warning is also important here. What you think about this?

[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/utils/_utils.py:667]: Yes lets have a warning

File: examples/embedding_model.py [hunk 2]
```diff
@@ -28,18 +27,20 @@ def mean_pooling(model_output, attention_mask):
 tokenizer = AutoTokenizer.from_pretrained("sentence-transformers/all-MiniLM-L6-v2")
 
 
-qeff_model = AutoModel.from_pretrained("sentence-transformers/all-MiniLM-L6-v2")
-qeff_model.compile(num_cores=14)
+# You can specify the pooling strategy either as a string (e.g., "mean") or by passing a custom pooling function.
+# If no pooling is specified, the model will return its default output (typically token embeddings).
+qeff_model = AutoModel.from_pretrained("sentence-transformers/all-MiniLM-L6-v2", pooling=max_pooling)
```
Comment (quic-rishinr) [inline examples/embedding_model.py:32]: Add one more commented line passing pooling as a string.

File: examples/embedding_model.py [hunk 3]
```diff
@@ -28,18 +27,20 @@ def mean_pooling(model_output, attention_mask):
...
+
+# Example: Using mean pooling by specifying it as a string.
+# This will return sentence embeddings computed using mean pooling.
+# qeff_model = AutoModel.from_pretrained("sentence-transformers/all-MiniLM-L6-v2", pooling="mean")
+
+# Here seq_len can be list seq_len or single int
+qeff_model.compile(num_cores=16, seq_len=[32, 64])
```
Comment (quic-rishinr) [inline examples/embedding_model.py:39]: Add one more commented line with seq_len as 32

File: QEfficient/transformers/models/modeling_auto.py [hunk 4]
```diff
@@ -271,7 +277,7 @@ def compile(
         onnx_path: Optional[str] = None,
         compile_dir: Optional[str] = None,
         *,
-        seq_len: int = 32,
+        seq_len: Union[int, List[int]] = 32,
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/modeling_auto.py:280]: Do we need to restrict the number of seq_len that can be added to a specialization? Thoughts?

File: tests/transformers/models/test_embedding_models.py [hunk 5]
```diff
@@ -11,17 +11,17 @@
...
 from QEfficient.utils._utils import create_json
 from QEfficient.utils.constants import Constants, QnnConstants
 
 embed_test_models = [
-    # model_name, architecture
-    "sentence-transformers/multi-qa-mpnet-base-cos-v1",  # MPNetForMaskedLM
-    "BAAI/bge-reranker-v2-m3",  # XLMRobertaForSequenceClassification
-    "BAAI/bge-small-en-v1.5",  # BertModel
+    {"model_name": "jinaai/jina-embeddings-v2-base-code", "pooling": "mean"},
+    {"model_name": "sentence-transformers/nli-bert-base-cls-pooling", "pooling": "cls"},
```
Comment (quic-rishinr) [inline tests/transformers/models/test_embedding_models.py:24]: do we need to have one test with pooling=None as well?

**Review by quic-rishinr — CHANGES_REQUESTED** [review]

File: examples/embedding_model.py (hunk 2 above)
Comment (quic-amitraj) [reply examples/embedding_model.py:32]: Its already there at line 36.

File: tests/transformers/models/test_embedding_models.py (hunk 5 above)
Comment (quic-amitraj) [reply tests/transformers/models/test_embedding_models.py:24]: Okay, is there any case you can think of?

File: QEfficient/transformers/models/pytorch_transforms.py [hunk 6]
```diff
@@ -487,3 +489,22 @@ class KVCacheModuleMethodMapperTransform(ModuleMethodMapperTransform):
...
+class PoolingTransform:
+    """
+    Apply a pooling transformation to the model. This transformation appends a pooling layer to the model, allowing for the reduction of spatial dimensions in the output.
+    The pooling layer can be configured to use different pooling methods, such as max pooling or average pooling.
+    """
+
+    @classmethod
+    def apply(cls, model: nn.Module, **kwargs) -> Tuple[nn.Module, bool]:
+        transformed = False
+        if kwargs.get("pooling") is not None:
```
[the remark this answers is not in the document]
Comment (quic-amitraj) [reply QEfficient/transformers/models/pytorch_transforms.py:503]: This is because we have not explicitly handled the pooling as a argument and instead of having if condition there I though it better to keep it here. If you want we can change init and take pooling as a argument. if its not none then we can apply this transform.

File: QEfficient/transformers/models/pytorch_transforms.py [hunk 7]
```diff
@@ -487,3 +489,22 @@ class KVCacheModuleMethodMapperTransform(ModuleMethodMapperTransform):
...
+            pooling = kwargs["pooling"]
+            pooling_method = (
+                POOLING_MAP[pooling] if isinstance(pooling, str) else validate_user_pooling_function(pooling)
```
[the remark this answers is not in the document]
Comment (quic-amitraj) [reply QEfficient/transformers/models/pytorch_transforms.py:506]: Nice catch, missed this. I will update.

**Review by quic-amitraj — COMMENTED** [review]
Comments addressed.

**Review by quic-amitraj — COMMENTED** [review]
Updated docstring in all module.

**Review by quic-hemagnih — APPROVED** [review]

**Review by quic-rishinr — APPROVED** [review]
